File: Problems/0600_0699/0606_Construct_String_from_Binary_Tree/Project_CS/Construct_String_from_Binary_Tree.cs
using System;

public class Solution
{
    public string Tree2str(TreeNode t)
    {
        if (t == null)
            return "";

        string resultStr = t.val.ToString();

        if (t.left == null && t.right == null)
            return resultStr;

        resultStr += "(" + Tree2str(t.left) + ")";
        if (t.right != null)
            resultStr += "(" + Tree2str(t.right) + ")";

        return resultStr;
    }

    public void Main(string args)
    {
        Console.WriteLine("args = " + args );
        string flds = args.Replace("\"", "").Replace("[", "").Replace("]", "").Trim();

        OperateTreeNode ope_t = new OperateTreeNode();
        TreeNode t = ope_t.CreateTreeNode(flds);
        Console.Write("t = \n" + ope_t.TreeToStaircaseString(t));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        string result = Tree2str(t);

        sw.Stop();
        Console.WriteLine("result = " + result);
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}

File: Problems/0600_0699/0606_Construct_String_from_Binary_Tree/Project_Python3/Construct_String_from_Binary_Tree.py
# coding: utf-8

import os
import sys
import time

from TreeNode.Codec import Codec
from TreeNode.TreeNode import TreeNode
from TreeNode.OperateTreeNode import OperateTreeNode

class Solution:
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if t is None:
            return ""

        resultStr = str(t.val)

        if t.left is None and t.right is None:
            return resultStr

        resultStr += "(" + self.tree2str(t.left) + ")"
        if t.right is not None:
            resultStr += "(" + self.tree2str(t.right) + ")"

        return resultStr

    def tree2str_2(self, t):
        if not t: return ''
        left = '({})'.format(self.tree2str(t.left)) if (t.left or t.right) else ''
        right = '({})'.format(self.tree2str(t.right)) if t.right else ''
        return '{}{}{}'.format(t.val, left, right)

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    flds = temp.replace(", ", ",").replace("\"", "").replace("[", "").replace("]", "").rstrip()

    ope_t = OperateTreeNode()
    t = ope_t.createTreeNode(flds.split(","))
    print("t = \n{0}".format(ope_t.treeToStaircaseString(t)))
#   print("t = {0}".format(ope_t.tree2str(t)))

    sl = Solution()
    time0 = time.time()
    result = sl.tree2str(t)

    time1 = time.time()

    print("result = {0}".format(result))
    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
